### PR TITLE
Add routed networks support to OVB

### PR DESCRIPTION
**Pull request by hjensas**
This change adds support to deploy routed networks:
  3x Provisioning networks
  2x Networks for each isolated net (internal_api, tenant, etc)

Neutron routers are configured to provide the routing between
the different networks.

A dhcprealy instance is created. It will relay DHCP requests
on networks provisioning_leaf1 and provisioning_leaf2 to the
undercloud ip addresses on the provisioning_leaf0 network.

The undercloud is connected to provision_leaf0.
Baremetal nodes are connected to provision_leaf1 and
provision_leaf2. By default 1 baremetal node is set up in
each leaf, the number of nodes are controlled by the
node_count_leaf1 and node_count_leaf2 parameters.

To deploy:
  bin/deploy.py \
    --quintupleo_routed \
    -e environments/quintupleo-routed.yaml
    -e ../overrides_env.yaml

A new option '--physical_network' is added to build-nodes-json.
When this option is used the nodes json uses the new (since Rocky)
schema to define ports, and includes the physical network as well
as the mac address when baremetal ports are created.

To create nodes_json for importing bm's in undercloud run the
build-nodes-json command twice using each leafs respective
prefixes and provision_net. For example:

  bin/build-nodes-json \
    --bmc_prefix bmc_leaf1 \
    --baremetal_prefix baremetal_leaf1 \
    --provision_net provision_leaf1 \
    --driver ipmi \
    --physical_network leaf1 \
    --nodes_json ../nodes_leaf1.json

  bin/build-nodes-json \
    --bmc_prefix bmc_leaf2 \
    --baremetal_prefix baremetal_leaf2 \
    --provision_net provision_leaf2 \
    --driver ipmi \
    --physical_network leaf2 \
    --nodes_json ../nodes_leaf2.json

### CODE REVIEW

File: environments/quintupleo-routed.yaml [hunk 1]
```diff
@@ -0,0 +1,68 @@
+parameter_defaults:
```
Comment (cybertron) [inline environments/quintupleo-routed.yaml]: Environments in ovb should be generated based on https://github.com/cybertron/openstack-virtual-baremetal/blob/master/sample-env-generator/environments.yaml

**Review by cybertron — COMMENTED** [review]
Okay, so this is a huge change but a great use for OVB.  It would be a nightmare setting this up by hand.

That said, I'm not sure I'm comfortable with the amount of near-duplication that's happening.  As far as I can tell, at a high level this is adding the following:

- support for additional networks
- support for a router and dhcp relay between those networks
- support for deploying multiple roles that connect to different sets of those networks

The only part of that which should require major changes is the router/dhcp relay.  OVB already has support for custom network templates (https://github.com/cybertron/openstack-virtual-baremetal/blob/master/templates/baremetal-networks-all.yaml) and multiple roles with custom configuration (https://github.com/cybertron/openstack-virtual-baremetal/blob/master/environments/base-role.yaml).  It looks to me like it should be possible to implement this with minimal changes for the latter two.  We would need to add handling of the new network names to the id and role functions (with the advent of composable networks that whole piece of functionality should be more flexible, but for this purpose I'm fine with continuing to hard-code network names) and the roles would need to have appropriate networking resource_registry entries added.  They would basically look like the resource_registry entries in https://github.com/cybertron/openstack-virtual-baremetal/blob/master/environments/all-networks-port-security.yaml  It looks like everything else should be inheritable from the base configuration.

The other advantage of implementing this as separate roles is the networking templates get simpler because we no longer need conditionals.  Each role would get configured with only the appropriate networks in the first place.  Also, doing this with roles _should_ eliminate the need to generate separate nodes.json files.  The role functionality names everything such that it can all be rolled up into one nodes.json, with optional separate nodes.jsons for the individual roles if desired.  For ci purposes I would think that's a useful thing.

So essentially what you would do is create two roles - the first (base) one would be responsible for deploying the undercloud, all of the networks/routers/relays, and the leaf1 overcloud nodes.  The second would just be responsible for the leaf2 overcloud nodes.

I realize that this is a pretty huge shift in direction, but this is exactly the kind of thing the roles functionality was designed to handle (multiple overcloud node types with potentially different network/disk/memory layouts) so I would strongly prefer to pursue it.  I am, of course, happy to help with that.  It's worth it to me to hopefully avoid maintaining another major deployment type going forward.

**Comment by cybertron** [comment]
Just to update: I started a new branch with some changes to enable this with roles: https://github.com/cybertron/openstack-virtual-baremetal/tree/routed-networks

So far it's just the physical_network changes and some abstraction of the undercloud network templates so they can be fully customized, including new networks.  I did tweak the physical_network change a bit to simplify it.  Let me know if what I did is wrong though.

It's also worth noting that this is technically backwards-incompatible.  That said, I think we're at a point where some of the really old behaviors and features need to be dropped, including the ones broken by these changes.  Probably the thing to do will be to cut a stable branch for anyone who might still be running against really old clouds and such, then merge this and the other changes I want to make to master.

Anyway, I'll continue integrating your changes into that branch and see if I can get it working.  I just wanted to get the initial work pushed for backup purposes and so you can take a look and see if it seems to be headed in the right direction.

**Comment by hjensas** [comment]
Thanks Ben! Sorry fpr not getting back sooner. I have been on PTO, and will continue to be on PTO for two more weeks. Will have a look at your new branch when I am back home.

**Comment by cybertron** [comment]
Yeah, no rush.  What's up there right now isn't all that close to complete, but hopefully this week I'll be able to start integrating your template changes and see if this is going to work.

**Pull request closed without merging**